### DISCUSSION TITLE
real, explicit serialization of loader responses

### DISCUSSION THREAD

**Discussion by rolandjohann** [remix-run/remix#6462] [Proposals]
The loader and action response payload properties will be serialized implicitly by `JSON.stringify` which results in a best guess for complex, non stdlib types and can't be reversed at e.g. `useLoaderData<typeof loader>()`.  There are several discussions of how problematic the current `SerializedFrom<T>` type is and that there are third party libs utilizing `superjson` or similar approaches.
https://github.com/remix-run/remix/issues/3931
https://github.com/remix-run/remix/issues/4626

At other languages/frameworks serialization is a more or less explicit process that can be setup as required.

I'd like to propose a change in the loader and action API so remix can handle the serialization and deserialization natively - if desired.

```typescript
interface LoaderFunction {
  (args: DataFunctionArgs): Promise<Response> | Response | Promise<AppData> | AppData;
}

interface SerDeLoader<T extends Record<string, unknown>, D = T> {
  load(args: LoaderArgs):
    | TypedResponse<T>
    | Promise<TypedResponse<T>>
    | TypedDeferredData<T>
    | Promise<TypedDeferredData<T>>;
  serialize(body: T): string;
  deserialize(body: unknown): D;
}

type Function = LoaderFunction | SerDeLoader;
```

With this the `LoaderFunction` will use an implementation utilizing default JSOn serialization/deserialization but it is open for others to plug in the ser/de of choice. the hoocks using the loaders then can be adapted as well to also utilize the deserialization of the `SerDeLoader` implementation.

**Comment by sergiodxa** [comment]
So you suggest changing from:
```ts
export async function loader(args: DataFunctionArgs) { ... }
```
to:
```ts
export let loader = {
  async loader(args) { ... },
  serialize(body) { ... },
  deserialize(body) { ... }
} satisfies SerDeLoader
```

I once proposed to have built-in support for superjson and a single place to configure extra objects for serialization/deserialization.

Right now, I think this is easier to do when you call `json` instead of the whole loader. Also I'm not completely sure you can remove the loader and serialize functions from the client bundle and keep deserialize so the app can parse back values.

@kiliman [remix-typedjson](https://github.com/kiliman/remix-typedjson) and @donavon [superjson-remix](https://github.com/donavon/superjson-remix) are a great example of how you can solve this when you create the response and in the useLoaderData hook.

In both libraries it's still not global but it's more granualar as it's per response and not the whole loader.

**Reply by rolandjohann** [reply]
@sergiodxa thank you for the detailed feedback!

The reason I came up with the `SerDeLoader` interface is that the hooks are then able to properly deserialize the responses and infer the types correctly. With this I didn't consider things like client bundling.

`remix-typedjson` and `superjson-remix` are great, but are not remix native. In addition, there are other possibilities to serialize/deserialize (e.g. [io-ts](https://github.com/gcanti/io-ts)), that won't add information to the deserialized output, like `superjson` does. `remix-typedjson` for example won't be able to handle non standard JS types, like `luxon datetime`, or such.

Now I recognize that an API change is nothing that I should propose, but a feature request 😅.

It would be great to be able to natively specify serialization/deserialization instead of relying to the implicit `JSON` serialization/deserialization.

> In both libraries it's still not global but it's more granualar as it's per response and not the whole loader.

This is the issue, IMHO. With this the dev is responsible to ensure that every response is deserialized correctly. At the loader usage via `useLoaderData` you won't know which response will be exactly returned and the dev is again responsible to ensure that the response is deserialized properly. Sure, the types are inferred somehow, but the serialization/deserialization needs to be implemented explicitly everywhere.

Having this issue solved natively would provide an integrated DX when dealing with loaders and actions. At the beginning of our project with remix it was not transparent to us that the loader responses will be serialized introducing implicit conversions of non native JSON types, because we assumed it is handled at the server at SSR time - so why should the types differ, when dealing with complex types? With `Date`s the issue came up that we need to deal serialization/deserialization manually. Now it's obvious, since we return `Response`s from the loader/action.

**Reply by kiliman** [reply]
Yes, I too wish that Remix provided a way to add our own serialize/deserialize, encode/decode function. That's one of the sticking points I have with adding support for `defer` in `remix-typedjson`. The `JSON.stringify` is done deep in the internals as it streams the response. We should be able to provide our own functions to replace the default `JSON.stringify`.

I'm not sure you need to modify the `loader/action` interface. You could expose those extensions in _remix.config.js_. It could use the same signature as `JSON.stringify/parse`

```js
interface JSON {
    /**
     * Converts a JavaScript Object Notation (JSON) string into an object.
     * @param text A valid JSON string.
     * @param reviver A function that transforms the results. This function is called for each member of the object.
     * If a member contains nested objects, the nested objects are transformed before the parent object is.
     */
    parse(text: string, reviver?: (this: any, key: string, value: any) => any): any;
    /**
     * Converts a JavaScript value to a JavaScript Object Notation (JSON) string.
     * @param value A JavaScript value, usually an object or array, to be converted.
     * @param replacer A function that transforms the results.
     * @param space Adds indentation, white space, and line break characters to the return-value JSON text to make it easier to read.
     */
    stringify(value: any, replacer?: (this: any, key: string, value: any) => any, space?: string | number): string;
    /**
     * Converts a JavaScript value to a JavaScript Object Notation (JSON) string.
     * @param value A JavaScript value, usually an object or array, to be converted.
     * @param replacer An array of strings and numbers that acts as an approved list for selecting the object properties that will be stringified.
     * @param space Adds indentation, white space, and line break characters to the return-value JSON text to make it easier to read.
     */
    stringify(value: any, replacer?: (number | string)[] | null, space?: string | number): string;
}

module.exports = {
  parse: myParser,
  stringify: myStringifier,
}
```
